New Twig Manual (Partial) (#1082)
contao/docs: docs/manual/layout/templates/twig/encoding.en.md

(and why you should favor output encoding) in this [OWASP article](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#rule-0-never-insert-untrusted-data-except-in-allowed-locations) about preventing Cross Site Scripting (XSS) attacks.

## Why you should care

The gist: You, as a template designer, have to decide how things should be output, because *you* know the context and which content you 
trust or not. The *exact same* data can be dangerous in one context and harmless in another:

With Twig we can be specific how a certain variable should be treated. Use the `|escape` or - short - `|e` 
[filter](https://twig.symfony.com/doc/3.x/filters/escape.html) for this:

```twig
<style>
  .box { background: {{ color|e('css') }} }
</style>

[…]

<div class="box">{{ color|e('html') }}</div>
```

{{% notice note %}}Hall
By default Twig encodes **all** variables. The chosen escaper strategy will depend on the template's file extension: your `.html.twig`
templates will automatically get the `|e('html')` treatment, so you could omit this part in the above example.
{{% /notice %}}

> 
> You can mark a whole section of a template to be escaped or not by using the [autoescape](https://twig.symfony.com/doc/3.x/tags/autoescape.html) tag.

Reviewer: I'd remove this because it's not relevant for users.

```suggestion

You can mark a whole section of a template to be escaped or not by using the [autoescape](https://twig.symfony.com/doc/3.x/tags/autoescape.html) tag.
```

## Trusted content

> If you want to intentionally output a variable that contains HTML, you must add the 
> [Filter](https://twig.symfony.com/doc/3.x/filters/raw.html) `|raw` to the variable. Remember that you should always use `|raw` only in 
> connection with trusted content. However, in connection with input via the TinyMCE editor in the backend, you can use the `|raw` filter in 
> this particular case.

Reviewer: ```suggestion
If you want to intentionally output something "as is" - like a variable, that contains HTML, you must add the
[Filter](https://twig.symfony.com/doc/3.x/filters/raw.html) `|raw` to the variable. Remember that you must **never** use `|raw` if you do not trust the data, otherwise this will create an XSS vulnerability.

A typical use case, where it is fine to use `|raw`, is back end content entered via the TinyMCE editor. Contao will already sanitize the user input (based on the sanitizer settings) before it passes the data to the template.
```

> Example: The content element of type "Text" contains the following entry `<p>My<br>Text</p>`:
> 
> ```twig
> {% extends "@Contao/content_element/text.html.twig" %}
> 
> {% block content %}
> 
>   {{ text }}
>   {# Encode #}
>   {# yields: "&lt;p&gt;My&lt;br&gt;Text&lt;/p&gt;" #}
> 
>   {{ text|raw }}
>   {# Do not encode #}
>   {# yields: "<p>My<br>Text</p>" #}
> 
> {% endblock %}
> ```

Reviewer: I tried to boil down this example as well:

```suggestion
In the following example, a context variable `text` has the content `<p>My<br>Text</p>`, while a variable `user_input` has the content `<script>alert('I am unsafe!');</script>`. Note, how adding `|raw` in the last example is dangerous, while it is fine in the second:

```twig
{{ text }}
{# outputs: "&lt;p&gt;My&lt;br&gt;Text&lt;/p&gt;" #}

{{ text|raw }}
{# outputs: "<p>My<br>Text</p>" #}

{{ user_input|raw }}
{# outputs the potentially dangerous: "<script>alert('I am unsafe!');</script>" #}
```
```

[filter](https://twig.symfony.com/doc/3.x/filters/escape.html) for this:

```twig
<style>
  .box { background: {{ color|e('css') }} }
</style>

[…]

<div class="box">{{ color|e('html') }}</div>
```

{{% notice note %}}
By default Twig encodes **all** variables. The chosen escaper
strategy will depend on the template's file extension: your `.html.twig`
templates will automatically get the `|e('html')` treatment, so you could omit
this part in the above example.
{{% /notice %}}

You can mark a whole section of a template to be escaped or not by using the 
[autoescape](https://twig.symfony.com/doc/3.x/tags/autoescape.html) tag.

## Trusted raw data

If you intentionally **do** want to output a variable containing raw HTML, like 
`<b>nice</b>`, you need to add the `|raw` [filter](https://twig.symfony.com/doc/3.x/filters/raw.html) to your variable which tells 
Twig to skip escaping this value. Otherwise `&lt;b&gt;nice&lt;/b&gt;` will be output, i.e. a text saying *&lt;b&gt;nice&lt;/b&gt;* and 
not a bold word <b>nice</b>. 

{{% notice warning %}}
Keep in mind, that you only ever add `|raw` to trusted input! Using `|raw` on 
anything else may result in severe XSS vulnerabilities!

> {{% /notice %}}

Reviewer: IMHO we should also explain how to use the `|insert_tag_raw` filter here. Because this will prevent people from doing `foo|insert_tag|raw` when they actually wanted to write `foo|insert_tag_raw`.

To prevent the question coming up in the first place, we should IMHO also point out, that the Contao back end sanitizes the text coming from a tinyMCE field for use in HTML and that - in this very case - it is safe to write `|raw`.

Author: Thought that we should avoid insertags in templates? 
And wouldn't we have to list all the filters then? 

We want to introduce practical Twig examples in a separate PR (Twig FAQ). Maybe that would be more suitable there then

Reviewer: There's a difference between using insert tags in the template (e.g. instead of a Twig filter/function or to insert something) and explicitly replacing insert tags in a piece of text.

```twig
{# Do not replace insert tags, encode #}
{{ text }}
e.g. yields "&lt;p&gt;my{{br}}text&lt;/p&gt;"

{# Do not replace insert tags, don't encode (= text raw): #}
{{ text|raw }}
e.g. yields "<p>my{{br}}text</p>"

{# Replace insert tags, encode everything: #}
{{ text|insert_tag }}
e.g. yields "&lt;p&gt;my&lt;br&gt;text&lt;/p&gt;"

{# Replace insert tags, but *only* encode the text around the insert tags (= insert tags raw): #}
{{ text|insert_tag_raw }}
e.g. yields "&lt;p&gt;my<br>text&lt;/p&gt;" (note the intact "<br>")

{# Replace insert tags, don'T encode (= all raw): #}
{{ text|insert_tag|raw }}
e.g. yields "<p><br></p>"
```

As you can see in the above example, only the template author can know if insert tags should be replaced, if they may contain HTML that will be output as such and if the text surrounding them may contain HTML that is output as such.

Author: Suggestion in last commit

Author: @m-vo But I am not sure if the examples describe the differences sufficiently and understandably?
